code/_onclick/drag_drop.dm

```diff
@@ -1,28 +1,42 @@
 /*
-	MouseDrop:
-
-	Called on the atom you're dragging.  In a lot of circumstances we want to use the
+	The below procs are called on the atom you're dragging.  In a lot of circumstances we want to use the
 	recieving object instead, so that's the default action.  This allows you to drag
 	almost anything into a trash can.
 */
 
-/atom/proc/CanMouseDrop(atom/over, var/mob/user = usr, var/incapacitation_flags)
-	if(!user || !over)
+/atom/MouseDrop(atom/over)
+	SHOULD_CALL_PARENT(TRUE)
+	if(!can_mouse_drop(over, usr) || !handle_mouse_drop(over, usr))
+		. = ..()
+
+/atom/proc/handle_mouse_drop(var/atom/over, var/mob/user)
+	addtimer(CALLBACK(over, /atom/proc/receive_mouse_drop, src, user), 0)
+	return TRUE
+
+// This proc checks if an atom can be mousedropped onto the target by the user.
+/atom/proc/can_mouse_drop(var/atom/over, var/mob/user = usr, var/incapacitation_flags = INCAPACITATION_DEFAULT)
+	SHOULD_CALL_PARENT(TRUE)
+	if(!istype(user) || !istype(over) ||QDELETED(user) || QDELETED(over) || QDELETED(src))
+		return FALSE
+	if(user.incapacitated(incapacitation_flags) || !CanPhysicallyInteract(user) || !user.check_dexterity(DEXTERITY_GRIP))
 		return FALSE
-	if(user.incapacitated(incapacitation_flags))
+	if(!check_mousedrop_adjacency(over, user))
 		return FALSE
-	if(!src.Adjacent(user) || !over.Adjacent(user))
-		return FALSE // should stop you from dragging through windows
 	return TRUE
 
-/atom/MouseDrop(atom/over)
-	if(!usr || !over) return
-	if(!Adjacent(usr) || !over.Adjacent(usr)) return // should stop you from dragging through windows
+/atom/proc/check_mousedrop_adjacency(var/atom/over, var/mob/user)
+	. = (Adjacent(user) && over.Adjacent(user))
 
-	spawn(0)
-		over.MouseDrop_T(src,usr)
-	return
+/mob/can_mouse_drop(var/atom/over, var/mob/user = usr, var/incapacitation_flags = INCAPACITATION_DEFAULT)
+	. = !anchored && ..()
 
-// Receive a mouse drop
-/atom/proc/MouseDrop_T(atom/dropping, mob/user)
-	return
+// Receive a mouse drop.
+// Returns true if the atom is valid for dropping further up the chain, false if the drop has been handled.
+/atom/proc/receive_mouse_drop(var/atom/dropping, var/mob/user)
+	SHOULD_CALL_PARENT(TRUE)
```

Comment:
I'm actually a bit hesitant about this. I realize you can unset it, but I'm not sure this should be the expected/default behavior. I'll trust your reading of the actual usage on it, though.

Reply:
It was set because the root receive_mouse_drop was doing a bunch of checking as commented below. Now that the checks and async are removed I can remove it.